core/src/com/google/inject/internal/aop/Enhancer.java

```diff
@@ -0,0 +1,421 @@
+ *     switch (index) {
+ *       case 0: {
+ *         return new HostClass$$EnhancerByGuice((InvocationHandler[]) context, ...);
+ *       }
+ *       case 1: {
+ *         return context.super.instanceMethod(...); // call original unenhanced method
+ *       }
+ *     }
+ *     return null;
+ *   }
+ *
+ *   // enhanced method
+ *   public final Object instanceMethod(...) {
+ *     // pack arguments and trigger the associated InterceptorStackCallback
+ *     return GUICE$HANDLERS[0].invoke(this, null, args);
+ *   }
+ *
+ *   // ...
+ * }
+ * </pre>
+ *
+ * @author mcculls@gmail.com (Stuart McCulloch)
+ */
+final class Enhancer extends AbstractGlueGenerator {
+
+  private static final String HANDLERS_NAME = "GUICE$HANDLERS";
+
+  private static final String HANDLERS_DESCRIPTOR = "[Ljava/lang/reflect/InvocationHandler;";
+
+  private static final String HANDLER_TYPE = Type.getInternalName(InvocationHandler.class);
+
+  private static final String HANDLER_ARRAY_TYPE = Type.getInternalName(InvocationHandler[].class);
+
+  private static final String INVOKERS_NAME = "GUICE$INVOKERS";
+
+  private static final String INVOKERS_DESCRIPTOR = "Ljava/lang/invoke/MethodHandle;";
+
+  private static final String CALLBACK_DESCRIPTOR =
+      "(Ljava/lang/Object;"
+          + "Ljava/lang/reflect/Method;"
+          + "[Ljava/lang/Object;)"
+          + "Ljava/lang/Object;";
+
+  // Describes the LambdaMetafactory.metafactory method arguments and return type
+  private static final String METAFACTORY_DESCRIPTOR =
+      "(Ljava/lang/invoke/MethodHandles$Lookup;"
+          + "Ljava/lang/String;"
+          + "Ljava/lang/invoke/MethodType;"
+          + "Ljava/lang/invoke/MethodType;"
+          + "Ljava/lang/invoke/MethodHandle;"
+          + "Ljava/lang/invoke/MethodType;)"
+          + "Ljava/lang/invoke/CallSite;";
+
+  private static final Type INDEX_TO_INVOKER_METHOD_TYPE =
+      Type.getMethodType("(I)Ljava/util/function/BiFunction;");
+
+  private static final Type RAW_INVOKER_METHOD_TYPE =
+      Type.getMethodType("(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;");
+
+  private static final Type INVOKER_METHOD_TYPE =
+      Type.getMethodType("(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;");
+
+  private final Map<Method, Method> bridgeDelegates;
+
+  private final String checkcastToProxy;
+
+  Enhancer(Class<?> hostClass, Map<Method, Method> bridgeDelegates) {
+    super(hostClass, ENHANCER_BY_GUICE_MARKER);
+    this.bridgeDelegates = bridgeDelegates;
+
+    // CHECKCAST(proxyName) fails when hosted anonymously; hostName works in that scenario
+    this.checkcastToProxy = ClassDefining.isAnonymousHost(hostClass) ? hostName : proxyName;
+  }
+
+  @Override
+  protected byte[] generateGlue(Collection<Executable> members) {
+    ClassWriter cw = new ClassWriter(COMPUTE_MAXS);
+
+    cw.visit(V1_8, PUBLIC | ACC_SUPER, proxyName, null, hostName, null);
```

Comment:
add a comment on why we're using v1_8 compatible classes (as opposed to anything older or newer)?

also: does this limit our ability to interact with host code that uses newer JDK features?

also: we use the constant in a bunch of places, should we extract to a shared constant so we can increment throughout the codebase more easily?  or is that not a meaningful thing to do?

Reply:
We use `V1_8` as a minimum because that's the base version that Guice currently supports. The bytecode generated for the fast-class and enhancer glue is simple and doesn't need anything more than Java8. This is because we're not copying the actual method code, only generating invokers that match the expected signature.

I'd prefer not to use a shared constant here because it makes it harder to see what level we're targeting. Also in the future we may decide we want to change the version for one of the cases, but not the others.

Reply:
774b89b